### PR TITLE
Tooltips for starrating (lp 1487855)

### PR DESCRIPTION
**Pull request by jclaveau**
This PR adds the support of the tooltips for starrating.  
I opened it on the 1.12 branch as i don't consider it as a new feature but a bug and it seems easy to review and check.

There is one last question I can't answer : For blind people, should I display the value that would be given if the left button of the mouse is clicked?

I just opened this bug to track it.
https://bugs.launchpad.net/mixxx/+bug/1487855

### CODE REVIEW

File: src/widget/wstarrating.h [hunk 1]
```diff
@@ -32,7 +32,9 @@ class WStarRating : public QWidget, public WBaseWidget {
     void mouseMoveEvent(QMouseEvent *event);
     void mouseReleaseEvent(QMouseEvent *event);
     void leaveEvent(QEvent *);
-    
+    virtual bool event(QEvent* pEvent);
+    void fillDebugTooltip(QStringList* debug);
```
Comment (ywwg) [inline src/widget/wstarrating.h]: I think this needs the virtual keyword because it's an override.

Comment (jclaveau) [inline src/widget/wstarrating.h]: When I add it, the tooltips are not customized anymore.

Here is the example of the pushbutton (without virtual) : https://github.com/mixxxdj/mixxx/blob/master/src/widget/wpushbutton.h#L77-L84

Comment (jclaveau) [inline src/widget/wstarrating.h]: Erratum : something with my build cache didn't work so the tooltips are correctly displayed even with virtual.

And in wslider composed it's a public slot :/ 
https://github.com/mixxxdj/mixxx/blob/master/src/widget/wslidercomposed.h#L52-L55

Comment (jclaveau) [inline src/widget/wstarrating.h]: And from here : https://en.wikipedia.org/wiki/Virtual_function
> In C++, virtual methods are declared by prepending the virtual keyword to the function's declaration in the base class. This modifier is inherited by all implementations of that method in derived classes, meaning that they can continue to over-ride each other and be late-bound

Should I still add it?

Comment (daschuer) [inline src/widget/wstarrating.h]: For documentation purpose, it is a good practice to repeat the virtual keyword in inherited classes.

Comment (ywwg) [inline src/widget/wstarrating.h]: in the master branch, which is c++11,  we'll also be adding the "override" keyword for super-extra confirmation

File: src/widget/wstarrating.cpp [hunk 2]
```diff
@@ -6,7 +6,8 @@
 #include "widget/wstarrating.h"
 
 WStarRating::WStarRating(QString group, QWidget* pParent)
-        : WBaseWidget(pParent),
+        : QWidget(pParent),
+          WBaseWidget(this),
```
Comment (daschuer) [inline src/widget/wstarrating.cpp]: You should inherit from WWidget, which is derived form QWidget and WBaseWidget

File: src/widget/wstarrating.cpp [hunk 3]
```diff
@@ -118,3 +119,20 @@ void WStarRating::mouseReleaseEvent(QMouseEvent*) {
 
     m_pCurrentTrack->setRating(m_starRating.starCount());
 }
+
+bool WStarRating::event(QEvent* pEvent) {
```
Comment (daschuer) [inline src/widget/wstarrating.cpp]: this is already done in WWidget

File: src/widget/wstarrating.h (hunk 1 above)
Comment (uklotzde) [inline src/widget/wstarrating.h]: C++11: "override" is enough in derived classes, you don't need to add "virtual" again. "override" is only allowed for virtual functions in derived classes. The compiler will complain if you add "override" and there is no corresponding virtual function with a matching signature in the base class.

Before switching to C++11 I added a trailing "/*override*/" comment to virtual functions in derived classes. This string can easily be replaced after switching to C++11 and you get the extra compiler check for free.

File: src/widget/wstarrating.cpp (hunk 2 above)
Comment (jclaveau) [inline src/widget/wstarrating.cpp]: This is what I did in the old PR https://github.com/mixxxdj/mixxx/pull/613 but @rryan noticed it and I thought I was wrong. Let's do it again! \o/

File: src/widget/wstarrating.h (hunk 1 above)
Comment (jclaveau) [inline src/widget/wstarrating.h]: It is done. I added "virtual" and  "/override/" as comment even if I've not been able to find this string somewhere else in the sources. Please, tell me if I'm wrong.

File: src/widget/wstarrating.cpp (hunk 2 above)
Comment (jclaveau) [inline src/widget/wstarrating.cpp]: It should be better now.

File: src/widget/wstarrating.cpp (hunk 3 above)
Comment (jclaveau) [inline src/widget/wstarrating.cpp]: Removed now.

File: src/widget/wstarrating.h (hunk 1 above)
Comment (uklotzde) [inline src/widget/wstarrating.h]: The master branch is already on C++11. All "override" comments have been 
replaced by the keyword.

**Comment by jclaveau** [comment]
All notes look addressed now.

**Comment by jclaveau** [comment]
This PR doesn't seem to arouse a lot of interest even if it's quite simple. Is there any chance it will be reviewed for 1.12 or should I recreate it for master?

**Comment by ywwg** [comment]
yeah master is a better place for this.  git rebase --onto master should do the trick, and then edit the PR's target.

**Comment by rryan** [comment]
Hello! This branch is against the 1.12 branch, which is now closed to any changes except bugfixes. Please re-open your PR against master if this PR is still relevant.

**Pull request closed without merging**